File: 2/1_Cuatrimestre/Programacion_y_Diseño_Orientado_a_Objetos/JavaFinal/NapakalakiJavaSinInterfaz/src/NapakalakiGame/Player.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package NapakalakiGame;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Random;
import GUI.Dice;
//import GUI.Dice;
/**
 *
 * @author Javi
 */
public class Player {
    static int MAXLEVEL=10;
    private String name;
    private int level;
    private boolean dead=true;
    private boolean canISteal=true;
    protected Player enemy;
    private BadConsequence pendingBadConsequence;
    private ArrayList <Treasure> hiddenTreasures=new ArrayList();
    private ArrayList <Treasure> visibleTreasures=new ArrayList();
    
    public Player(String name){
        this.name=name;
        this.level=1;
    }
    public Player(Player p){
        this.name=p.name;
        this.level=p.level;
        this.dead=p.dead;
        this.canISteal=p.canISteal;
        this.enemy=p.enemy;
        this.pendingBadConsequence=p.pendingBadConsequence;
        this.hiddenTreasures=p.hiddenTreasures;
        this.visibleTreasures=p.visibleTreasures;
    }
    
    public String getName(){
        return name;
    }
    public Player getEnemy(){
        return enemy;
    }
    private void bringToLife(){
        dead=false;
    }
    public int getCombatLevel(){
        int nivel_combate=this.level;
        for(Treasure t : visibleTreasures){
            nivel_combate=nivel_combate+t.getBonus();
        }
        return nivel_combate;
    }
    private void incrementLevels(int l){
        level=level+l;
    }
    private void decrementLevels(int l){
        level=level-l;
        if(level<1){
            level=1;
        }
    }
    private void setPendingBadConsequence(BadConsequence b){
        pendingBadConsequence=b;
    }
    private void applyPrize(Monster m){
        int nLevels=m.getLevelsGained();
        this.incrementLevels(nLevels);
        int nTreasures=m.getTreasuresGained();
        if(nTreasures>0){
            CardDealer dealer=CardDealer.getInstance();
            for(int i=0; i<nTreasures;i++){
                hiddenTreasures.add(dealer.nextTreasure());
            }
        }
    }
    private void applyBadConsequence(Monster m){
        int nLevels=m.getBadConsequence().getLevels();
        this.decrementLevels(nLevels);
        this.pendingBadConsequence=m.getBadConsequence().adjustToFitTreasureLists(visibleTreasures, hiddenTreasures);
        this.setPendingBadConsequence(this.pendingBadConsequence);
    }
    private boolean canMakeTreasureVisible(Treasure t){
        boolean podemos=true;
        int NumeroOneHand=0;
        boolean BothHandEquipada=false;
        for(Treasure t2: visibleTreasures){
            if(t2.getType()==TreasureKind.ONEHAND){
                NumeroOneHand++;
            }
            if(t2.getType()==TreasureKind.BOTHHANDS){
                BothHandEquipada=true;
            }
            if(t2.getType()==t.getType() && t2.getType()!=TreasureKind.ONEHAND && t2.getType()!=TreasureKind.BOTHHANDS){
                podemos=false;
            }
        }
        if( (t.getType()==TreasureKind.ONEHAND && NumeroOneHand>=2) || (t.getType()==TreasureKind.BOTHHANDS && BothHandEquipada) || (t.getType()==TreasureKind.ONEHAND && BothHandEquipada) || (t.getType()==TreasureKind.BOTHHANDS && NumeroOneHand!=0) ){
            podemos=false;
        }
        return podemos;
    }
    private int howManyVisibleTreasures(TreasureKind tKind){
        int cont=0;
        for(Treasure t : visibleTreasures){
            if(t.getType()==tKind){
                cont++;
            }
        }
        return cont;
    }
    private void diefNoTreasures(){
        if(hiddenTreasures.isEmpty() && visibleTreasures.isEmpty()){
            dead=true;
        }
    }
    public boolean isDead(){
        return dead;
    }
    public ArrayList<Treasure> getHiddenTreasures(){
        return hiddenTreasures;
    }
    public ArrayList <Treasure> getVisibleTreasures(){
        return visibleTreasures;
    }
    public CombatResult combat(Monster m){
        int myLevel= this.getCombatLevel();
        int monsterLevel=this.getOponentLevel(m);
        if(myLevel>monsterLevel){
            this.applyPrize(m);
            if(level>=MAXLEVEL){
                return CombatResult.WINGAME;
            }
            else{
                return CombatResult.WIN;
            }
        }
        else{
            this.applyBadConsequence(m);
            if(this.shouldConvert()){
                return CombatResult.LOSEANDCONVERT;
            }
            else{
                return CombatResult.LOSE;
            }
        }
        
    }
    public void makeTreasureVisible(Treasure t){
        boolean canI= this.canMakeTreasureVisible(t);
        if(canI){
            visibleTreasures.add(t);
            hiddenTreasures.remove(t);
        }
    }
    public void discardVisibleTreasure(Treasure t){
        visibleTreasures.remove(t);
        if( (pendingBadConsequence!=null) && (!pendingBadConsequence.isEmpty())){
            pendingBadConsequence.substractVisibleTreasure(t);
        }
        this.diefNoTreasures();
    }
    public void discardHiddenTreasure(Treasure t){
        hiddenTreasures.remove(t);
        if( (pendingBadConsequence!=null) && (!pendingBadConsequence.isEmpty()) ){
            pendingBadConsequence.substratHiddenTreasure(t);
        }
        this.diefNoTreasures();
    }
    public boolean validState(){
        if(pendingBadConsequence==null){
            return true;
        }
        else if(pendingBadConsequence.isEmpty() && hiddenTreasures.size()<=4){
            return true;
        }
        else{
            return false;
        }
    }
    public void initTreasures(){
        CardDealer dealer=CardDealer.getInstance();
        Dice dice=Dice.getInstance();
        this.bringToLife();;
        Treasure treasure=dealer.nextTreasure();
        hiddenTreasures.add(treasure);
        int number=dice.nextNumber();
        if(number>1){
            treasure=dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
        if(number==6){
            treasure=dealer.nextTreasure();
            hiddenTreasures.add(treasure);
        }
    }
    public int getLevels(){
        return level;
    }
    public Treasure stealTreasure(){
        Treasure t=null;
        if(this.canISteal && enemy.canYouGiveMeATreasure()){
            t=enemy.giveMeATreasure();
            hiddenTreasures.add(t);
            this.haveStolen();
        }
        return t;
    }
    public void setEenemy(Player enemy){
        this.enemy=enemy;
    }
    private Treasure giveMeATreasure(){
        if(canYouGiveMeATreasure()){
            Random r=new Random();
            int i=r.nextInt(hiddenTreasures.size());
            Treasure t;
            t=hiddenTreasures.get(i);
            hiddenTreasures.remove(i);
            return t;
        }
        else return null;
    }
    public boolean canISteal(){
        return canISteal;
    }
    private boolean canYouGiveMeATreasure(){
        if(hiddenTreasures.isEmpty()){
            return false;
        }
        else{
            return true;
        }
    }
    private void haveStolen(){
        canISteal=false;
    }
    public void discardAllTreasures(){
        ArrayList<Treasure>vt=new ArrayList(visibleTreasures);
        ArrayList<Treasure>hd=new ArrayList(hiddenTreasures);
        for(Treasure t: vt){
            this.discardVisibleTreasure(t);
            pendingBadConsequence.substractVisibleTreasure(t);
        }
        for(Treasure t: hd){
            this.discardHiddenTreasure(t);
            pendingBadConsequence.substratHiddenTreasure(t);
        }
    }
    public String toString(){
        System.out.println( name+"  level->"+level+" levelCombat-->"+this.getCombatLevel()+"  dead->"+dead+" enemigo-->"+enemy.getName());
        if(this.pendingBadConsequence!=null){
            return "PendingBad----->" +this.pendingBadConsequence.toString();
        }
        else return " vacio";
    }
    
    protected int getOponentLevel(Monster m){
        return m.getCombatLevel();
    }
    protected boolean shouldConvert(){
       Dice dice=Dice.getInstance();
       if(dice.nextNumber()==1){
           return true;
       }
       else{
           return false;
       }
    }
    public BadConsequence getPendingBadConsequence(){
        if(pendingBadConsequence!=null){
        return pendingBadConsequence;}
        else return null;
    }
}

File: 2/1_Cuatrimestre/PDOO/RubyFinal/NapakalakiRuby/pkg/NapakalakiRuby-0.0.1/lib/Player.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
require_relative 'BadConsequence.rb'
require_relative 'Treasure.rb'
require_relative 'TreasureKind.rb'
require_relative 'Monster.rb'
require_relative 'CombatResult.rb'
require_relative 'Dice.rb'
module NapakalakiGame
  class Player
    @@MAXLEVEL=10
    def self.MAXLEVEL
      @@MAXLEVEL
    end
    attr_accessor :name, :level, :hiddenTreasures, :visibleTreasures, :dead, :pendingBadConsequence, :enemy
    def initialize(name)
      @name=name
      @dead=true
      @level=1
      @visibleTreasures=Array.new
      @hiddenTreasures=Array.new
    end
    def getHiddenTreasures()
      return @hiddenTreasures
    end
    def getVisibleTreasures()
      return @visibleTreasures
    end
    def isDead()
      if(@dead)then
        return true
      else
        return false
      end
    end
    def combat(m)
      myLevel=getCombatLevel
      monsterLevel=m.combatlevel
        if(myLevel>monsterLevel)then
          applyPrize(m)
          if @level>=10 then
            return CombatResult::WINGAME
          else
            return CombatResult::WIN
          end
        else
          applyBadConsequence(m)
          return CombatResult::LOSE
        end
    end
    def makeTreasureVisible(t)
        canI=self.canMakeTreasureVisible(t)
        if(canI)then
          @visibleTreasures<<t
          @hiddenTreasures.delete(t)
        end
    end
    def discardVisibleTreasure(t)
      @visibleTreasures.delete(t)
      if( (@pendingBadConsequence!=nil) && (!@pendingBadConsequence.isEmpty) )then
        @pendingBadConsequence.substractVisibleTreasure(t)
      end
      self.diefNoTreasures()
    end
    def discardHiddenTreasure(t)
      @hiddenTreasures.delete(t)
      if( (@pendingBadConsequence!=nil) && (!@pendingBadConsequence.isEmpty) )then
        @pendingBadConsequence.substractHiddenTreasure(t)
      end
      self.diefNoTreasures()
    end
    def validState()
      if (@pendingBadConsequence==nil) then
        return true
      elsif(@pendingBadConsequence.isEmpty && @hiddenTreasures.size<=4)then
        return true
      else
        return false
      end
    end
    def initTreasures()
      dealer=CardDealer.instance
      dice=Dice.instance
      bringToLife
      treasure=dealer.nextTreasure
      @hiddenTreasures<<treasure
      number=dice.nextNumber
      if(number>1)then
        treasure=dealer.nextTreasure
        @hiddenTreasures<<treasure
      end
      if(number==6)then
        treasure=dealer.nextTreasure
        @hiddenTreasures<<treasure
      end
    end
    def stealTreasure()
      t=Treasure.new
      t=nil
      if(self.canISteal && @enemy.canYouGiveMeATreasure())then
        t=@enemy.giveMeATreasure
        @hiddenTreasures<<t
        self.haveStolen
      end
      return t
    end
    def setEnemy(enemy)
      @enemy=enemy
    end
    def canISteal()
      return @canISteal
    end
    def discardAllTreasures()
      vt=@visibleTreasures
      hd=@hiddenTreasures
      vt.each { |t|  
        discardVisibleTreasure(t)
      }
      hd.each { |t|  
        discardHiddenTreasure(t)
      }
    end
    def bringToLife()
      @dead=false
    end
    def getCombatLevel()
      nc=0
      nc=@level+nc
      visibleTreasures.each { |vt|  
        nc=nc+vt.Bonus
      }
      return nc
    end
    def incrementLevels(l)
      @level=@level+l
    end
    def decrementLevels(l)
      @level=@level-l
      if (@level<1)then
        @level=1
      end
    end
    def setPendingBadConsequence(b)
      @pendingBadConsequence=b
    end
    def applyPrize(m)
      nLevels=m.getLevelsGained
      self.incrementLevels(nLevels)
      nTreasures=m.getTreasuresGained
      if(nTreasures>0)then
        dealer=CardDealer.getInstance
        for i in(0.. nTreasures)
          @hiddenTreasures<<dealer.nextTreasure()
        end
      end
    end
    def applyBadConsequence(m)
      nLevels=m.badConsequence.levels
      decrementLevels(nLevels)
      @pendingBadConsequence=m.badConsequence.adjustToFitTreasureLists(@visibleTreasures, @hiddenTreasures)
      setPendingBadConsequence(@pendingBadConsequence)
    end
    def canMakeTreasureVisible(t)
      podemos=true
      @visibleTreasures.each { |t2|  
        if(t2.type==t.type)then
          podemos=false
        end
        if( (t2.type==TreasureKind::ONEHAND || t2.type==TreasureKind::BOTHHANDS) && (t.type==TreasureKind::ONEHAND || t.type==TreasureKind::BOTHHANDS) )then
          podemos=false
        end
      }
      return podemos
    end
    def howManyVisibleTreasures(tKind)
      count=0
      @visibleTreasures.each { |i|  
        if(i.type==tKind)then
          count=count+1
        end
      }
      return count
    end
    def diefNoTreasures()
      if(hiddenTreasures.empty? && visibleTreasures.empty?)then
        dead=true
      end
    end
    def giveMeATreasure()
      if(canYouGiveMeATreasure)then
        i=rand(@hiddenTreasures.size)
        t=@hiddenTreasures.at(i)
        @hiddenTreasures.delete_at(i)
        return t
      else
        return nil
      end
    end
    def canYouGiveMeATreasure()
      if(hiddenTreasures.empty?)
        return false
      else 
        return true
      end
    end
    def haveStolen()
      @canISteal=false
    end
    def to_s() 
      "#{@name}"
    end
    
   
  end
end
